src/ucp/core/ucp_request.c

```diff
@@ -396,23 +396,36 @@ static ucp_md_map_t ucp_request_get_invalidation_map(ucp_ep_h ep)
     return inv_map;
 }
 
-int ucp_request_memh_invalidate(ucp_request_t *req, ucp_mem_h memh,
-                                ucs_status_t status)
+/* Get the contig memh from the request basing on the proto version */
+static ucp_mem_h ucp_request_get_contig_memh(ucp_request_t *req) {
+    if (req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED) {
+        ucs_assertv(req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG,
+                    "wrong dt_class : %s",
+                    ucp_datatype_class_names[req->send.state.dt_iter.dt_class]);
+        return req->send.state.dt_iter.type.contig.memh;
+    }
+
+    ucs_assertv(UCP_DT_IS_CONTIG(req->send.datatype),
+                "wrong datatype=0x%"PRIx64, req->send.datatype);
+    return req->send.state.dt.dt.contig.memh;
+}
+
+int ucp_request_memh_invalidate(ucp_request_t *req, ucs_status_t status)
 {
     ucp_ep_h ep                      = req->send.ep;
     ucp_err_handling_mode_t err_mode = ucp_ep_config(ep)->key.err_mode;
     ucp_worker_h worker              = ep->worker;
     ucp_context_h context            = worker->context;
+    ucp_mem_h memh                   = ucp_request_get_contig_memh(req);
```

Comment:
it seems we again assume the datatype is config - otherwise there is assert.
can't we call this function also when dt is not contain?

Reply:
Right, thanks, moved later as discussed.